File: Python/Data Structures/Segment Tree.py
import math
# Time: 
# Construction: O(nlogn)
# Query: O(logn)
# Space: O(n)

class SegmentTree():
    def __init__(self, arr, f, base):
        n = len(arr)
        self.n = n 
        self.tree = [0]*2*n
        self.f = f
        self.base = base
        self.build(arr)   
    
    def build(self, arr):
        n = self.n
        for i in range(n): 
            self.tree[n+i] = arr[i]; 
        
        for i in range(n-1, 0, -1): 
            self.tree[i] = self.f(self.tree[i << 1], self.tree[i << 1 | 1])

    def update(self, index, val): 
        n = self.n
        self.tree[index + n] = val
        index = index + n;         
        i = index
        
        while i > 1:
            self.tree[i >> 1] = self.f(self.tree[i], self.tree[i^1])
            i >>= 1; 
    
    def query(self, l, r):
        l += self.n; r += self.n+1
        res = self.base

        while l < r:
            if (l & 1):
                res = self.f(res, self.tree[l])
                l += 1
            if (r & 1):
                r -= 1
                res = self.f(res, self.tree[r])
            l >>= 1; r >>= 1
            
        return res; 

a = [1, 3, 5, 1, -1]; 
# st = SegmentTree(a, f=lambda x, y: max(x, y), base=-math.inf)
# st = SegmentTree(a, f=lambda x, y: min(x, y), base=math.inf)
st = SegmentTree(a, f=lambda x, y: x+y, base=0)
print(st.query(1, 4))

File: C++/Data Structures/Segment Tree.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1e9 + 7;
const long long INF = 1e9;

template<typename T>
struct RangeUpdate {
    vector<T> t;
    
    RangeUpdate(int n){
        t.assign(4*n, 0);
        return;
    }
 
    void build(vector<T>& a, int v, int tl, int tr) {
        if (tl == tr) {
            t[v] = a[tl];
        } else {
            int tm = (tl + tr) / 2;
            build(a, v*2, tl, tm);
            build(a, v*2+1, tm+1, tr);
            t[v] = 0;
        }
    }

    void update(int v, int tl, int tr, int l, int r, T add) {
        if (l > r) return;
        if (l == tl && r == tr) {t[v] += add;} 
        else {
            int tm = (tl + tr) / 2;
            update(v*2, tl, tm, l, min(r, tm), add);
            update(v*2+1, tm+1, tr, max(l, tm+1), r, add);
        }
    }

    T get(int v, int tl, int tr, int pos) {
        if (tl == tr)
            return t[v];
        int tm = (tl + tr) / 2;
        if (pos <= tm)
            return t[v] + get(v*2, tl, tm, pos);
        else
            return t[v] + get(v*2+1, tm+1, tr, pos);
    }
};
 


template<typename T>
struct SegmentTree{
    int n;
    vector<T> t;
    
    SegmentTree(int n){
        this->n = n;
        t.assign(4*n, 0);
        return;
    }
 
    T f(T a, T b){
        return a+b;
    }
 
    void build(vector<T>& a, int v, int tl, int tr){
        if(tl == tr){t[v] = a[tl];} 
        else{
            int tm = tl+(tr-tl)/2;
            build(a, v*2, tl, tm);
            build(a, v*2+1, tm+1, tr);
            t[v] = f(t[v*2], t[v*2+1]);
        }
    }

    void build(vector<T>& a){
        build(a, 1, 0, n-1);
    }
 
    T query(int v, int tl, int tr, int l, int r) {
        if(l > r){return 0;}
        if(l == tl && r == tr){return t[v];}
        int tm = tl+(tr-tl)/2;
        return f(query(v*2, tl, tm, l, min(r, tm)), query(v*2+1, tm+1, tr, max(l, tm+1), r));
    }

    T query(int l, int r) {
        return query(1, 0, n-1, l, r);
    }
 
    void update(int v, int tl, int tr, int pos, T new_val){
        if (tl == tr){t[v] = new_val;} 
        else{
            int tm = tl+(tr-tl)/2;
            if (pos <= tm){
                update(v*2, tl, tm, pos, new_val);
            }
            else{
                update(v*2+1, tm+1, tr, pos, new_val);
            }
            t[v] = f(t[v*2], t[v*2+1]);
        }
    }
    void update(int pos, T new_val){
        return update(1, 0, n-1, pos, new_val);
    }
};


int main(){
    vector<int> arr= {3,2,1};
    int n = arr.size();
    SegmentTree<int> st = SegmentTree<int>(n);
    st.build(arr);
    int res = st.query(0, n-1);
    cout << res;
    return 0;
}